equational-reasoning.md

```diff
@@ -0,0 +1,25 @@
+Equational reasoning
+====================
+
+**Tags**: *substitution model*, *referential transparency*, *purity*, *side effect*
+
+In terms of programming we can say that a function definition defines the scope where different expressions are evaluated in order to eventually return a result. This result defines the return type of that function.
+
+For developers it seems very convenient to be able to `reason` about a function (i.e. understand what the function does ) just by checking the value that it returns. This is to say that a function's return value represents the transformation of its internal state by potentially using other arguments as its parameters.
+
+However what a function does is not always related to the value that it returns and this is the case when a function has side effects. This is when we say that a function is not `pure`. By side effect we mean any kind of interaction that a function has with the outside word (i.e. outside of its scope ). This can be IO operations, altering state of external variables etc. As a result, the only way to reason about what the function does is to deep dive into the definition of the function and monitor what happens when a side effect operation happens.
+
+This breaks `referential transparency` meaning that what a function does is no more represented by the value that it returns. When referential transparency exists we can replace a function / expression with the value that returns thus being able to reason about our programs just by substituting the expressions we have with the values they return ( similar to the evaluation of mathematical equations ). This is what we call the `substitution model`. In other words we can say that `referential transparency` is the key that enables a natural way of reasoning ( according to mathematics ) based on `substitution model` by substituting equals with equals ( i.e. variables with/or referents with values). This is what we call in other words `equational reasoning`.
```

Comment:
` what a function does is no more represented by the value that it returns` this is good, but you are focusing on the return type/value, remember that reasoning about a function ALSO involves thinking about the input type/value. So for example, this phrase can be improved by saying: `This breaks referential transparency, meaning that what a function does is no more represented only by the values that it takes and that it returns.`

Reply:
Good point, I agree.